example/extensions/lib_subgraph/subgraph_lib.cc

```diff
@@ -0,0 +1,253 @@
+      // null is an input data to the subgraph, add to data storage
+      data.push_back(inputs[input_cnt++]);
+    } else if (op.compare("log") == 0) {
+      // get input tensor based on node ID inputs from data storage
+      MXTensor &input = data[node_inputs.list[0].list[0].num];
+      // create temporary storage
+      MXTensor tmp(malloc(input.size()*4), input.shape, input.dtype);
+      // save allocated ptr to free later
+      to_free.push_back(tmp.data_ptr);
+      // execute log operator
+      myLog(input,tmp);
+      // add output tensor to data storage
+      data.push_back(tmp);
+    } else if (op.compare("exp") == 0) {
+      // get input tensor based on node ID inputs from data storage
+      MXTensor &input = data[node_inputs.list[0].list[0].num];
+      // create temporary storage
+      MXTensor tmp(malloc(input.size()*4), input.shape, input.dtype);
+      // save allocated ptr to free later
+      to_free.push_back(tmp.data_ptr);
+      // execute exp operator 
+      myExp(input,tmp);
+      // add output tensor to data storage
+      data.push_back(tmp);
+    } else {
+      std::cout << "Error! Unsupported op '" << op << "' found in myExecutor";
+      return MX_FAIL;
+    }
+  }
+  
+  // get list of outputs from subgraph
+  JsonVal heads = json_val.map[JsonVal("heads")];
+  // copy all operator results to outputs of subgraph
+  for (int j = 0; j < heads.list.size(); j++) {
+    // get computed result
+    MXTensor &result = data[heads.list[0].list[0].num];
+    // get output tensor to pass to MX
+    MXTensor &out = outputs[j];
+    float *out_data = out.data<float>();
+    float *res_data = result.data<float>();
+    // loop and copy data
+    for (int i = 0; i < result.size(); i++) {
+      out_data[i] = res_data[i];
+    }
+  }
+
+  // free allocated temporary storage
+  for (void* ptr : to_free) {
+    free(ptr);
+  }
+  
+  return MX_SUCCESS;
+}
+
+class MyStatefulOp : public CustomStatefulOp {
+ public:
+  explicit MyStatefulOp(std::string sym) : subgraph_sym(sym) {}
+
+  MXReturnValue Forward(std::vector<MXTensor> inputs,
+                        std::vector<MXTensor> outputs,
+                        OpResource op_res) {
+    return myExecutor(inputs, outputs, subgraph_sym);
+  }
+
+ private:
+  std::string subgraph_sym;
+};
+
+MXReturnValue createOpState(std::map<std::string, std::string> attrs,
+                            CustomStatefulOp** op_inst) {
+  std::string serialized_subgraph = "[empty]";
+  // MXNet subgraph is stored as Symbol in operator node attrs subgraphs field
+  // custom subgraph is stored as json string in custom operator attrs map entry
+  if (attrs.count(SUBGRAPH_SYM_JSON)) {
+    // user can now parse json and run other custom ops inside subgraph
+    serialized_subgraph = attrs[SUBGRAPH_SYM_JSON];
+  }
+  *op_inst = new MyStatefulOp(serialized_subgraph);
+  std::cout << "Info: stateful operator created" << std::endl;
+  return MX_SUCCESS;
+}
+
+REGISTER_OP(_custom_subgraph_op)
```

Comment:
A new registration function? What's the difference with previous NNVM_REGISTER_OP and MXNET_REGISTER_OP?

Reply:
This is not a new one, its the same one from the customOp PR:
https://github.com/apache/incubator-mxnet/blob/f86a8d10074f983b990ef420b96d664e911ee1bf/include/mxnet/lib_api.h#L643-L646

The single header file that library creates must include does not include any MXNet headers at all. This creates separation and make creating a library easier.
https://github.com/apache/incubator-mxnet/blob/f86a8d10074f983b990ef420b96d664e911ee1bf/include/mxnet/lib_api.h#L28-L39